Fixes #2: Simplify mpy.List class

akhilpandey95 opened the pull request:
[MAJOR]
- added an argument to List for supporting ordinal vs nominal data
- added an argument to List for naming the List Variable expression
- Updated docstring comment style for List before the __init__ magic function
- added asserts in freeze() method to check if the idx's are of type list

[TRIVIAL]
- removed type hinting for the BaseSampler class in metalgpy/_sample.py

Signed-off-by: akhilpandey95 <akhilpandey.akella@gmail.com>

Commit:
Issue #2: Simplify mpy.List class

[MAJOR]
- added an argument to List for supporting ordinal vs nominal data
- added an argument to List for naming the List Variable expression
- Updated docstring comment style for List before the __init__ magic function
- added asserts in freeze() method to check if the idx's are of type list

[TRIVIAL]
- removed type hinting for the BaseSampler class in metalgpy/_sample.py

Signed-off-by: akhilpandey95 <akhilpandey.akella@gmail.com>

akhilpandey95 commented:
@Deathn0t Please squash and merge the PR and not merge.

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
-        k (int, optional): the number of values selected in values. Defaults to None.
-        replace (bool, optional): draw from values with replacement. Defaults to False.
-        invariant (bool, optional): values is permutation invariant (e.g., a list of same object types). Defaults to True.
+    ----------
```
For the documentation follow the same norm as deephyper, [here is an example](https://github.com/deephyper/deephyper/blob/develop/deephyper/search/hps/_cbo.py#L26)

akhilpandey95 replied:
Sure.

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
+                 ordered=False,
+                 k=None,
+                 replace=False,
+                 invariant=False,
```
`k, replace, invariant` needs to be removed

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
         self._invariant = invariant
+        self._ordered = ordered
 
+    # add dunder method to update string representation of the object
```
unnecessary comment related to Python official doc (remove)

akhilpandey95 replied:
Okay.

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
+    # add dunder method to update string representation of the object
     def __repr__(self) -> str:
+        """
+        cat_ordinal = mpy.List(values=["low", "medium", "high"], ordered=True, name="cat_ordinal")
```
code examples can be presented with `.. code-block::` [example here](https://github.com/deephyper/deephyper/blob/develop/deephyper/problem/_neuralarchitecture.py#L107) from Sphinx or preceded by `>>> ...` ([example here](https://github.com/deephyper/deephyper/blob/develop/deephyper/problem/_hyperparameter.py#L165), interpreter fashion)

akhilpandey95 replied:
Sure

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
+            raise ValueError(
+                f"index of List should be int or iterable but is {item} with type '{type(item)}'!")
 
+    # mangle the length method to the class object
```
"mangle"?

akhilpandey95 replied:
I thought it was a double underscore and felt you were name mangling, but it is just a single underscore, so its just a private method for python. Ignore the comment, I shall remove it.

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
+        # return length of given Variable expression
         return len(self._values)
 
+    # add dunder method to check operator equality
```
unnecessary comment, however the definition of equality could be mentioned as a method comment `"""..."""`

akhilpandey95 replied:
I think it could be removed.

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -495,58 +496,99 @@ def choices(self):
     def freeze(self, choice: dict):
-
-        if self._k is None:  # "replace" is ignored (only 1 sample is drawn)
+        # "replace" is ignored (only 1 sample is drawn)
```
the `freeze` method needs to be re-written to remove `k, invariant, replace` features and just keep `ordered` and the fact that nested variables can happen

akhilpandey95 replied:
Okay

Deathn0t commented on metalgpy/_sample.py:
```diff
@@ -36,11 +36,11 @@ def list_type(list_exp):
 
 class BaseSampler:
-    def __init__(self, exp, distributions: List[tuple]=None, rng:np.random.RandomState=None):
+    def __init__(self, exp, distributions, rng:np.random.RandomState=None):
```
the `distributions` parameter can `None` by default and `dict` otherwise, `distributions: dict=None`. In the `None` case, we will introspect the `exp` to collect the default `_dist`.

akhilpandey95 replied:
I'll go with `distributions: dict=None`

Deathn0t commented on tests/test_var.py:
```diff
@@ -26,37 +26,39 @@ def test_List(self):
         assert x1 == x2
 
+        cat_ordinal = mpy.List(values=["low", "medium", "high"], name="cat_ordinal")
+        cat_nominal = mpy.List(values=["red", "blue", "green", "yellow"], name="cat_nominal")
```
these 2 tests do not check much... what needs to be checked/ensured? where is `ordered`, use the `assert` statement

akhilpandey95 replied:
This was a mistake. Forgot to add assert. I shall add it. Apologies

Deathn0t reviewed (commented):
I left some comments, please take them into consideration.

akhilpandey95 committed:
Issue #2: Simplify mpy.List class

- rolled back the old docstring style comments beneath the List class
definition
- removed k, invariant, and replace arguments passed to the List class
- removed unnecessary comments pertaining to methods in the List class
- added asserts to check the validitiy of the ordered argument for the
List class inside tests/test_var
- updated freeze() method inside List() to ensure k, invariant, and
replace are no longer considered while freezing expression choices
- commented out the _sample() method in the list class

[TRIVIAL]
- updated type hinting for the BaseSampler class in metalgpy/_sample.py

Signed-off-by: Akhil Akella <aakella@gpu3.lcrc.anl.gov>

Deathn0t commented on metalgpy/_expression.py:
```diff
@@ -496,66 +496,30 @@ def choices(self):
-    Variable Expression
-        mpy.List
+    Args:
+        values (iterable): an interable with possible values.
```
should be a `list` not any iterable
